File: internal/handlers/profile_comment_handler.go
package handlers

import (
	"context"
	"net/http"
	"regexp"
	"strings"
	"time"

	"chat-backend/internal/database"
	"chat-backend/internal/models"

	"github.com/gin-gonic/gin"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
)

const (
	TargetTypePhone      = "phone"
	TargetTypeCarNumber  = "car_number"
	TargetTypePersonName = "person_name"
)

func normalizeTargetValue(s string) string {
	return strings.TrimSpace(strings.ReplaceAll(s, " ", ""))
}

type ProfileCommentHandler struct {
	db *database.Database
}

func NewProfileCommentHandler(db *database.Database) *ProfileCommentHandler {
	return &ProfileCommentHandler{db: db}
}

type CreateProfileCommentRequest struct {
	TargetUserID string `json:"target_user_id"` // optional when target_type+target_value used
	TargetType   string `json:"target_type"`    // "phone" | "car_number" | "person_name"
	TargetValue  string `json:"target_value"`   // phone, car plate, or person name
	Text         string `json:"text" binding:"required"`
}

// CreateProfileCommentByPhoneRequest: guest/public comment. Supports phone (notify+deletable by owner) or car_number/person_name (search-only).
type CreateProfileCommentByPhoneRequest struct {
	PhoneNumber string `json:"phone_number"` // legacy: when target_type is empty, treated as phone
	TargetType  string `json:"target_type"` // "phone" | "car_number" | "person_name"
	TargetValue string `json:"target_value"`
	Text        string `json:"text" binding:"required"`
}

// CreateProfileCommentByPhone creates a comment (guest). For "phone": user must exist, notification sent, owner can delete. For "car_number"/"person_name": only searchable.
func (h *ProfileCommentHandler) CreateProfileCommentByPhone(c *gin.Context) {
	var req CreateProfileCommentByPhoneRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	req.TargetType = strings.TrimSpace(strings.ToLower(req.TargetType))
	if req.TargetType == "" {
		req.TargetType = TargetTypePhone
	}
	if req.TargetType != TargetTypePhone && req.TargetType != TargetTypeCarNumber && req.TargetType != TargetTypePersonName {
		c.JSON(http.StatusBadRequest, gin.H{"error": "target_type must be phone, car_number, or person_name"})
		return
	}

	var targetValue string
	if req.TargetType == TargetTypePhone {
		targetValue = normalizeTargetValue(req.PhoneNumber)
		if targetValue == "" {
			targetValue = normalizeTargetValue(req.TargetValue)
		}
		if targetValue == "" {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid phone number"})
			return
		}
	} else {
		targetValue = strings.TrimSpace(req.TargetValue)
		if targetValue == "" {
			c.JSON(http.StatusBadRequest, gin.H{"error": "target_value required for car_number and person_name"})
			return
		}
	}

	comment := models.ProfileComment{
		ID:          primitive.NewObjectID(),
		TargetType:  req.TargetType,
		TargetValue: targetValue,
		CommenterID: primitive.NilObjectID,
		Text:        req.Text,
		CreatedAt:   time.Now(),
	}

	if req.TargetType == TargetTypePhone {
		var targetUser models.User
		err := h.db.MongoDB.Collection("users").FindOne(
			context.Background(),
			bson.M{"phone_number": targetValue},
		).Decode(&targetUser)
		if err != nil {
			c.JSON(http.StatusNotFound, gin.H{"error": "User not found with this phone number"})
			return
		}
		comment.TargetUserID = &targetUser.ID
		if _, err := h.db.MongoDB.Collection("profile_comments").InsertOne(context.Background(), comment); err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create comment"})
			return
		}
		notification := models.Notification{
			ID:        primitive.NewObjectID(),
			UserID:    targetUser.ID,
			Type:      "profile_comment",
			Title:     "Yeni şərh",
			Body:      "Nömrənizə yeni şərh yazıldı",
			Data: map[string]interface{}{
				"comment_id": comment.ID.Hex(),
				"phone":      targetUser.PhoneNumber,
			},
			CreatedAt: time.Now(),
		}
		h.db.MongoDB.Collection("notifications").InsertOne(context.Background(), notification)
	} else {
		comment.TargetUserID = nil
		if _, err := h.db.MongoDB.Collection("profile_comments").InsertOne(context.Background(), comment); err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create comment"})
			return
		}
	}

	c.JSON(http.StatusCreated, gin.H{
		"id":           comment.ID,
		"target_type":  comment.TargetType,
		"target_value": comment.TargetValue,
		"text":         comment.Text,
		"created_at":   comment.CreatedAt,
	})
}

// CreateProfileComment creates a comment (authenticated). Supports target_user_id (phone) or target_type+target_value (phone/car_number/person_name).
func (h *ProfileCommentHandler) CreateProfileComment(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	var req CreateProfileCommentRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	req.TargetType = strings.TrimSpace(strings.ToLower(req.TargetType))
	if req.TargetType == "" && req.TargetUserID != "" {
		req.TargetType = TargetTypePhone
	}
	if req.TargetType != "" && req.TargetType != TargetTypePhone && req.TargetType != TargetTypeCarNumber && req.TargetType != TargetTypePersonName {
		c.JSON(http.StatusBadRequest, gin.H{"error": "target_type must be phone, car_number, or person_name"})
		return
	}

	var targetID primitive.ObjectID
	usePhone := req.TargetType == TargetTypePhone || req.TargetUserID != ""
	if req.TargetUserID != "" {
		var err error
		targetID, err = primitive.ObjectIDFromHex(req.TargetUserID)
		if err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid target_user_id"})
			return
		}
	} else if req.TargetType == TargetTypePhone && req.TargetValue != "" {
		phone := normalizeTargetValue(req.TargetValue)
		var targetUser models.User
		err := h.db.MongoDB.Collection("users").FindOne(
			context.Background(),
			bson.M{"phone_number": phone},
		).Decode(&targetUser)
		if err != nil {
			c.JSON(http.StatusNotFound, gin.H{"error": "User not found with this phone number"})
			return
		}
		targetID = targetUser.ID
	}

	targetValue := strings.TrimSpace(req.TargetValue)
	if usePhone {
		if targetID == userIDObj {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Cannot comment on yourself"})
			return
		}
		comment := models.ProfileComment{
			ID:           primitive.NewObjectID(),
			TargetType:   TargetTypePhone,
			TargetValue:  targetValue,
			TargetUserID: &targetID,
			CommenterID:  userIDObj,
			Text:         req.Text,
			CreatedAt:    time.Now(),
		}
		if targetValue == "" {
			var u models.User
			if h.db.MongoDB.Collection("users").FindOne(context.Background(), bson.M{"_id": targetID}).Decode(&u) == nil {
				comment.TargetValue = u.PhoneNumber
			}
		}
		if _, err := h.db.MongoDB.Collection("profile_comments").InsertOne(context.Background(), comment); err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create comment"})
			return
		}
		var targetUser models.User
		if h.db.MongoDB.Collection("users").FindOne(context.Background(), bson.M{"_id": targetID}).Decode(&targetUser) == nil {
			notification := models.Notification{
				ID:        primitive.NewObjectID(),
				UserID:    targetID,
				Type:      "profile_comment",
				Title:     "Yeni şərh",
				Body:      "Nömrənizə yeni şərh yazıldı",
				Data: map[string]interface{}{
					"comment_id": comment.ID.Hex(),
					"phone":      targetUser.PhoneNumber,
				},
				CreatedAt: time.Now(),
			}
			h.db.MongoDB.Collection("notifications").InsertOne(context.Background(), notification)
		}
		c.JSON(http.StatusCreated, gin.H{
			"id":             comment.ID,
			"target_user_id": targetID.Hex(),
			"target_type":    comment.TargetType,
			"text":           comment.Text,
			"created_at":     comment.CreatedAt,
		})
		return
	}

	// car_number or person_name
	if req.TargetType == "" || (req.TargetType != TargetTypeCarNumber && req.TargetType != TargetTypePersonName) {
		c.JSON(http.StatusBadRequest, gin.H{"error": "target_type and target_value required (car_number or person_name)"})
		return
	}
	if targetValue == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "target_value required"})
		return
	}
	comment := models.ProfileComment{
		ID:           primitive.NewObjectID(),
		TargetType:   req.TargetType,
		TargetValue:  targetValue,
		TargetUserID: nil,
		CommenterID:  userIDObj,
		Text:         req.Text,
		CreatedAt:    time.Now(),
	}
	if _, err := h.db.MongoDB.Collection("profile_comments").InsertOne(context.Background(), comment); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create comment"})
		return
	}
	c.JSON(http.StatusCreated, gin.H{
		"id":           comment.ID,
		"target_type":  comment.TargetType,
		"target_value": comment.TargetValue,
		"text":         comment.Text,
		"created_at":   comment.CreatedAt,
	})
}

// GetProfileComments returns comments about a user. Author is never exposed.
// Can query by target_user_id or phone_number
func (h *ProfileCommentHandler) GetProfileComments(c *gin.Context) {
	targetIDStr := c.Query("target_user_id")
	phoneNumber := c.Query("phone_number")
	
	var targetID primitive.ObjectID
	var err error
	
	if targetIDStr != "" {
		targetID, err = primitive.ObjectIDFromHex(targetIDStr)
		if err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid target_user_id"})
			return
		}
	} else if phoneNumber != "" {
		// Find user by phone number
		var targetUser models.User
		err = h.db.MongoDB.Collection("users").FindOne(
			context.Background(),
			bson.M{"phone_number": phoneNumber},
		).Decode(&targetUser)
		if err != nil {
			c.JSON(http.StatusNotFound, gin.H{"error": "User not found with this phone number"})
			return
		}
		targetID = targetUser.ID
	} else {
		c.JSON(http.StatusBadRequest, gin.H{"error": "target_user_id or phone_number required"})
		return
	}

	cursor, err := h.db.MongoDB.Collection("profile_comments").Find(
		context.Background(),
		bson.M{"target_user_id": targetID},
	)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch comments"})
		return
	}
	defer cursor.Close(context.Background())

	var comments []models.ProfileComment
	if err := cursor.All(context.Background(), &comments); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to decode comments"})
		return
	}

	// Return without commenter_id (anonymous)
	out := make([]map[string]interface{}, 0, len(comments))
	for _, pc := range comments {
		out = append(out, map[string]interface{}{
			"id":            pc.ID,
			"text":          pc.Text,
			"like_count":    pc.LikeCount,
			"dislike_count": pc.DislikeCount,
			"created_at":    pc.CreatedAt,
		})
	}
	c.JSON(http.StatusOK, out)
}

// ReplyToProfileComment lets the profile owner start a chat with the commenter (commenter stays anonymous).
func (h *ProfileCommentHandler) ReplyToProfileComment(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	commentIDStr := c.Param("comment_id")
	commentID, err := primitive.ObjectIDFromHex(commentIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid comment_id"})
		return
	}

	var comment models.ProfileComment
	err = h.db.MongoDB.Collection("profile_comments").FindOne(
		context.Background(),
		bson.M{"_id": commentID},
	).Decode(&comment)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Comment not found"})
		return
	}

	// Only the target user (profile owner) can reply; only phone-type comments have target_user_id
	if comment.TargetUserID == nil || *comment.TargetUserID != userIDObj {
		c.JSON(http.StatusForbidden, gin.H{"error": "Only the profile owner can reply to this comment"})
		return
	}

	commenterID := comment.CommenterID
	if commenterID == primitive.NilObjectID {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Cannot reply to a comment from a guest"})
		return
	}
	if commenterID == userIDObj {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Cannot reply to your own comment"})
		return
	}

	// Find or create direct chat between profile owner and commenter; make chat anonymous FROM commenter so owner sees "Anonymous"
	members := []primitive.ObjectID{userIDObj, commenterID}
	var chat models.Chat
	err = h.db.MongoDB.Collection("chats").FindOne(
		context.Background(),
		bson.M{
			"type": "direct",
			"$and": []bson.M{
				{"members": bson.M{"$all": members}},
				{"members": bson.M{"$size": 2}},
			},
		},
	).Decode(&chat)
	if err != nil {
		// Create new chat; commenter is anonymous to profile owner
		chat = models.Chat{
			ID:                  primitive.NewObjectID(),
			Type:                "direct",
			Members:             members,
			AnonymousFromUserID: &commenterID,
			CreatedAt:           time.Now(),
			UpdatedAt:           time.Now(),
		}
		_, err = h.db.MongoDB.Collection("chats").InsertOne(context.Background(), chat)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create chat"})
			return
		}
	}

	c.JSON(http.StatusOK, gin.H{
		"chat_id": chat.ID.Hex(),
		"message": "Chat created; you can message the commenter (they appear as Anonymous)",
	})
}

// SearchProfileComments returns comments whose target_value matches the query (any type). Used to find car_number/person_name comments.
func (h *ProfileCommentHandler) SearchProfileComments(c *gin.Context) {
	q := strings.TrimSpace(c.Query("q"))
	if q == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "q required"})
		return
	}
	// Also find by phone: if q is a registered phone, include comments for that user
	var targetUserID *primitive.ObjectID
	var targetUser models.User
	if err := h.db.MongoDB.Collection("users").FindOne(
		context.Background(),
		bson.M{"phone_number": normalizeTargetValue(q)},
	).Decode(&targetUser); err == nil {
		targetUserID = &targetUser.ID
	}
	escaped := regexp.QuoteMeta(q)
	pattern := ".*" + escaped + ".*"
	orConditions := []bson.M{
		{"target_value": bson.M{"$regex": pattern, "$options": "i"}},
	}
	if targetUserID != nil {
		orConditions = append(orConditions, bson.M{"target_user_id": targetUserID})
	}
	filter := bson.M{"$or": orConditions}
	cursor, err := h.db.MongoDB.Collection("profile_comments").Find(
		context.Background(),
		filter,
	)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to search comments"})
		return
	}
	defer cursor.Close(context.Background())
	var comments []models.ProfileComment
	if err := cursor.All(context.Background(), &comments); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to decode comments"})
		return
	}
	out := make([]map[string]interface{}, 0, len(comments))
	for _, c := range comments {
		out = append(out, map[string]interface{}{
			"id":           c.ID,
			"target_type":  c.TargetType,
			"target_value": c.TargetValue,
			"text":         c.Text,
			"created_at":   c.CreatedAt,
		})
	}
	c.JSON(http.StatusOK, out)
}

// DeleteProfileComment allows the target user (number owner) to delete only phone-type comments about them.
func (h *ProfileCommentHandler) DeleteProfileComment(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)
	commentIDStr := c.Param("comment_id")
	commentID, err := primitive.ObjectIDFromHex(commentIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid comment_id"})
		return
	}
	var comment models.ProfileComment
	err = h.db.MongoDB.Collection("profile_comments").FindOne(
		context.Background(),
		bson.M{"_id": commentID},
	).Decode(&comment)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Comment not found"})
		return
	}
	if comment.TargetUserID == nil || *comment.TargetUserID != userIDObj {
		c.JSON(http.StatusForbidden, gin.H{"error": "Only the number owner can delete this comment"})
		return
	}
	_, err = h.db.MongoDB.Collection("profile_comments").DeleteOne(
		context.Background(),
		bson.M{"_id": commentID},
	)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to delete comment"})
		return
	}
	c.JSON(http.StatusOK, gin.H{"message": "Comment deleted"})
}

File: src/handlers/profileCommentHandler.js
'use strict';

const { ObjectId } = require('mongodb');
const { getDB } = require('../database');

// ─── Constants ────────────────────────────────────────────────────────────────

const TARGET_TYPE_PHONE = 'phone';
const TARGET_TYPE_CAR_NUMBER = 'car_number';
const TARGET_TYPE_PERSON_NAME = 'person_name';

const VALID_TARGET_TYPES = [TARGET_TYPE_PHONE, TARGET_TYPE_CAR_NUMBER, TARGET_TYPE_PERSON_NAME];

// ─── Helpers ──────────────────────────────────────────────────────────────────

function normalizeTargetType(t) {
  if (!t || !VALID_TARGET_TYPES.includes(t)) return TARGET_TYPE_PHONE;
  return t;
}

// ─── Handlers ─────────────────────────────────────────────────────────────────

/**
 * Public route — no auth required.
 */
async function createProfileCommentByPhone(req, res) {
  try {
    const db = getDB();
    const { phone_number, target_type, target_value, text } = req.body || {};

    if (!text) return res.status(400).json({ error: 'text is required' });
    if (!target_value && !phone_number) {
      return res.status(400).json({ error: 'target_value or phone_number is required' });
    }

    const normalizedType = normalizeTargetType(target_type);
    const resolvedTargetValue = target_value || phone_number;

    const now = new Date();
    const commentDoc = {
      target_type: normalizedType,
      target_value: resolvedTargetValue,
      text,
      like_count: 0,
      dislike_count: 0,
      created_at: now,
      updated_at: now,
    };

    // If phone type: try to find the user
    if (normalizedType === TARGET_TYPE_PHONE) {
      const phoneToLookup = phone_number || resolvedTargetValue;
      const targetUser = await db.collection('users').findOne({ phone_number: phoneToLookup });
      if (targetUser) {
        commentDoc.target_user_id = targetUser._id.toString();

        // Insert notification for the target user
        await db.collection('notifications').insertOne({
          user_id: targetUser._id.toString(),
          type: 'profile_comment',
          message: 'Someone left a comment on your profile',
          created_at: now,
        });
      }
    }

    const insertResult = await db.collection('profile_comments').insertOne(commentDoc);
    commentDoc._id = insertResult.insertedId;

    return res.status(201).json({ ...commentDoc, id: insertResult.insertedId.toString() });
  } catch (err) {
    console.error('createProfileCommentByPhone error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

/**
 * Auth-required route.
 */
async function createProfileComment(req, res) {
  try {
    const db = getDB();
    const userId = req.userId;
    const { target_user_id, target_type, target_value, text } = req.body || {};

    if (!text) return res.status(400).json({ error: 'text is required' });

    const normalizedType = normalizeTargetType(target_type);

    const now = new Date();
    const commentDoc = {
      commenter_id: userId,
      target_type: normalizedType,
      target_value: target_value || null,
      text,
      like_count: 0,
      dislike_count: 0,
      created_at: now,
      updated_at: now,
    };

    if (normalizedType === TARGET_TYPE_PHONE) {
      // Resolve target user by phone or target_user_id
      let targetUser = null;
      if (target_user_id) {
        try {
          targetUser = await db.collection('users').findOne({ _id: new ObjectId(target_user_id) });
        } catch (_) {}
      } else if (target_value) {
        targetUser = await db.collection('users').findOne({ phone_number: target_value });
      }

      if (targetUser) {
        commentDoc.target_user_id = targetUser._id.toString();

        // Insert notification
        await db.collection('notifications').insertOne({
          user_id: targetUser._id.toString(),
          type: 'profile_comment',
          message: 'Someone left a comment on your profile',
          created_at: now,
        });
      }
    } else {
      // car_number or person_name: no target_user_id
      if (target_user_id) commentDoc.target_user_id = target_user_id.toString();
    }

    const insertResult = await db.collection('profile_comments').insertOne(commentDoc);
    commentDoc._id = insertResult.insertedId;

    return res.status(201).json({ ...commentDoc, id: insertResult.insertedId.toString() });
  } catch (err) {
    console.error('createProfileComment error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function getProfileComments(req, res) {
  try {
    const db = getDB();
    const { target_user_id, phone_number } = req.query;

    const filter = {};
    if (target_user_id) {
      filter.target_user_id = target_user_id;
    } else if (phone_number) {
      filter.target_value = phone_number;
    } else {
      return res.status(400).json({ error: 'target_user_id or phone_number query param is required' });
    }

    const comments = await db.collection('profile_comments').find(filter)
      .sort({ created_at: -1 })
      .toArray();

    // Return without commenter_id (anonymous)
    return res.json(
      comments.map(c => ({
        id: c._id.toString(),
        text: c.text,
        like_count: c.like_count || 0,
        dislike_count: c.dislike_count || 0,
        target_type: c.target_type,
        target_value: c.target_value,
        created_at: c.created_at,
      }))
    );
  } catch (err) {
    console.error('getProfileComments error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function replyToProfileComment(req, res) {
  try {
    const db = getDB();
    const userId = req.userId;
    const commentIdStr = req.params.comment_id;

    let commentObjId;
    try {
      commentObjId = new ObjectId(commentIdStr);
    } catch (_) {
      return res.status(400).json({ error: 'Invalid comment ID' });
    }

    const comment = await db.collection('profile_comments').findOne({ _id: commentObjId });
    if (!comment) return res.status(404).json({ error: 'Comment not found' });

    // Only the target user can reply
    if (!comment.target_user_id || comment.target_user_id !== userId) {
      return res.status(403).json({ error: 'Only the target user can reply to this comment' });
    }

    const commenterIdStr = comment.commenter_id;
    if (!commenterIdStr) {
      return res.status(400).json({ error: 'No commenter to reply to' });
    }

    const now = new Date();

    // Find or create direct chat between target user and commenter (with anonymous_from_user_id)
    let chat = await db.collection('chats').findOne({
      type: 'direct',
      members: { $all: [userId, commenterIdStr] },
      anonymous_from_user_id: commenterIdStr,
    });

    if (!chat) {
      const chatDoc = {
        type: 'direct',
        members: [userId, commenterIdStr],
        anonymous_from_user_id: commenterIdStr,
        created_at: now,
        updated_at: now,
      };
      const chatInsert = await db.collection('chats').insertOne(chatDoc);
      chatDoc._id = chatInsert.insertedId;
      chat = chatDoc;
    }

    const chatIdStr = chat._id.toString();

    return res.json({
      chat_id: chatIdStr,
      message: 'You can now message the commenter anonymously',
    });
  } catch (err) {
    console.error('replyToProfileComment error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function searchProfileComments(req, res) {
  try {
    const db = getDB();
    const { q } = req.query;

    if (!q) return res.status(400).json({ error: 'q query param is required' });

    // Try to find a user with this phone number
    const matchedUser = await db.collection('users').findOne({ phone_number: q });

    const filter = {
      $or: [
        { target_value: { $regex: q, $options: 'i' } },
      ],
    };

    if (matchedUser) {
      filter.$or.push({ target_user_id: matchedUser._id.toString() });
    }

    const comments = await db.collection('profile_comments').find(filter)
      .sort({ created_at: -1 })
      .limit(50)
      .toArray();

    return res.json(
      comments.map(c => ({
        id: c._id.toString(),
        target_type: c.target_type,
        target_value: c.target_value,
        text: c.text,
        created_at: c.created_at,
      }))
    );
  } catch (err) {
    console.error('searchProfileComments error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function deleteProfileComment(req, res) {
  try {
    const db = getDB();
    const userId = req.userId;
    const commentIdStr = req.params.comment_id;

    let commentObjId;
    try {
      commentObjId = new ObjectId(commentIdStr);
    } catch (_) {
      return res.status(400).json({ error: 'Invalid comment ID' });
    }

    const comment = await db.collection('profile_comments').findOne({ _id: commentObjId });
    if (!comment) return res.status(404).json({ error: 'Comment not found' });

    // Only the target_user_id owner can delete
    if (!comment.target_user_id || comment.target_user_id !== userId) {
      return res.status(403).json({ error: 'Only the target user can delete this comment' });
    }

    await db.collection('profile_comments').deleteOne({ _id: commentObjId });

    return res.json({ message: 'Comment deleted' });
  } catch (err) {
    console.error('deleteProfileComment error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

module.exports = {
  createProfileCommentByPhone,
  createProfileComment,
  getProfileComments,
  replyToProfileComment,
  searchProfileComments,
  deleteProfileComment,
  TARGET_TYPE_PHONE,
  TARGET_TYPE_CAR_NUMBER,
  TARGET_TYPE_PERSON_NAME,
};
